store/plugin-commands-store-search/src/findHash.ts

```diff
@@ -0,0 +1,82 @@
+import path from 'path'
+import fs from 'fs'
+import chalk from 'chalk'
+
+import { type Config } from '@pnpm/config'
+import { PnpmError } from '@pnpm/error'
+import { getStorePath } from '@pnpm/store-path'
+
+import renderHelp from 'render-help'
+
+export const PACKAGE_INFO_CLR = chalk.greenBright
+export const INDEX_PATH_CLR = chalk.hex('#078487')
+
+export const commandNames = ['find-hash']
+
+export const rcOptionsTypes = cliOptionsTypes
+
+export function cliOptionsTypes () {
+  return {}
+}
+
+export function help () {
+  return renderHelp({
+    description:
+      'Print the contained packages according to the hash.',
+    descriptionLists: [],
+    usages: ['pnpm find-hash <hash>'],
+  })
+}
+
+export type findHashCommandOptions = Pick<Config, 'storeDir' | 'pnpmHomeDir'>
+export interface findHashResult {
+  name: string
+  version: string
+  indexPath: string
+}
+
+export async function handler (opts: findHashCommandOptions, params: string[]) {
+  if (!params || params.length === 0) {
+    throw new PnpmError('MISSING_HASH', '`pnpm find-hash` requires the hash')
+  }
+
+  const hash = params[0]
+  const storeDir = await getStorePath({
+    pkgRoot: process.cwd(),
+    storePath: opts.storeDir,
+    pnpmHomeDir: opts.pnpmHomeDir,
+  })
+  const cafsDir = path.join(storeDir, 'files')
+  const cafsChildrenDirs = fs.readdirSync(cafsDir).filter(dirName => fs.statSync(`${cafsDir}/${dirName}`).isDirectory())
```

Comment:
Use readdirSync with the withFileTypes option set to true. In that case you won't have to run statSync.